src/main/kotlin/no/nav/syfo/application/Authentication.kt

```diff
@@ -4,28 +4,34 @@ import com.auth0.jwk.JwkProvider
 import io.ktor.application.Application
 import io.ktor.application.install
 import io.ktor.auth.Authentication
+import io.ktor.auth.Principal
 import io.ktor.auth.UserIdPrincipal
 import io.ktor.auth.basic
+import io.ktor.auth.jwt.JWTCredential
 import io.ktor.auth.jwt.JWTPrincipal
 import io.ktor.auth.jwt.jwt
 import net.logstash.logback.argument.StructuredArguments
+import no.nav.syfo.Environment
 import no.nav.syfo.VaultSecrets
 import no.nav.syfo.log
 
-fun Application.setupAuth(vaultSecrets: VaultSecrets, jwkProvider: JwkProvider, issuer: String) {
+fun Application.setupAuth(vaultSecrets: VaultSecrets, jwkProvider: JwkProvider, issuer: String, env: Environment) {
     install(Authentication) {
         jwt(name = "jwt") {
             verifier(jwkProvider, issuer)
             validate { credentials ->
-                if (!credentials.payload.audience.contains(vaultSecrets.loginserviceClientId)) {
-                    log.warn(
-                        "Auth: Unexpected audience for jwt {}, {}",
-                        StructuredArguments.keyValue("issuer", credentials.payload.issuer),
-                        StructuredArguments.keyValue("audience", credentials.payload.audience)
-                    )
-                    null
-                } else {
-                    JWTPrincipal(credentials.payload)
+                when {
+                    hasLoginserviceClientIdAudience(credentials, vaultSecrets) -> JWTPrincipal(credentials.payload)
+                    else -> unauthorized(credentials)
+                }
+            }
+        }
+        jwt(name = "rerun") {
```

Comment:
Dette ble veldig ryddig og fint! Jeg er litt usikker på om det er samme issuer for selvbetjeningstokenene og de som appene fra fss får, det kan være verdt å sjekke.

Reply:
godt poeng!